pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,636 @@
+	}
+
+	return nil
+}
+
+func (a *actuator) registerAsSeed(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	// Check seed spec and create or update the seed secrets if needed
+	err := a.prepareSeedSpec(ctx, &managedSeed.Spec.SeedTemplate.Spec, managedSeed, shoot)
+	if err != nil {
+		return err
+	}
+
+	// Create or update the seed
+	return a.createOrUpdateSeed(ctx, managedSeed)
+}
+
+func (a *actuator) unregisterAsSeed(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed) error {
+	// Ensure the seed is deleted
+	if err := a.ensureSeedDeleted(ctx, managedSeed); err != nil {
+		return err
+	}
+
+	// Delete seed secrets if needed
+	return a.unprepareSeedSpec(ctx, &managedSeed.Spec.SeedTemplate.Spec, managedSeed)
+}
+
+func (a *actuator) createOrUpdateSeed(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed) error {
+	seed := &gardencorev1beta1.Seed{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: managedSeed.Name,
+		},
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, a.gardenClient.Client(), seed, func() error {
+		seed.OwnerReferences = []metav1.OwnerReference{
+			*metav1.NewControllerRef(managedSeed, seedmanagementv1alpha1.SchemeGroupVersion.WithKind("ManagedSeed")),
+		}
+		seed.Labels = utils.MergeStringMaps(managedSeed.Spec.SeedTemplate.Labels, map[string]string{
+			v1beta1constants.GardenRole: v1beta1constants.GardenRoleSeed,
+		})
+		seed.Annotations = managedSeed.Spec.SeedTemplate.Annotations
+		seed.Spec = managedSeed.Spec.SeedTemplate.Spec
+		return nil
+	})
+	return err
+}
+
+func (a *actuator) ensureSeedDeleted(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed) error {
+	// Delete the seed
+	seed := &gardencorev1beta1.Seed{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: managedSeed.Name,
+		},
+	}
+	if err := a.gardenClient.Client().Delete(ctx, seed); client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// Check if there are any remaining objects associated with the seed
+	if err := a.checkSeedAssociations(ctx, managedSeed.Name); err != nil {
+		return err
+	}
+
+	// Check if the seed still exists
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(managedSeed.Name), &gardencorev1beta1.Seed{}); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	return fmt.Errorf("seed %q still exists", managedSeed.Name)
+}
+
+func (a *actuator) deployGardenlet(ctx context.Context, shootClient kubernetes.Interface, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	// Decode gardenlet configuration
+	gardenletConfig, err := helper.DecodeGardenletConfigurationExternal(managedSeed.Spec.Gardenlet.Config, false)
+	if err != nil {
+		return err
+	}
+
+	// Check seed spec and create or update the seed secrets if needed
+	if err := a.prepareSeedSpec(ctx, &gardenletConfig.SeedConfig.SeedTemplate.Spec, managedSeed, shoot); err != nil {
```

Comment:
`helper.DecodeGardenletConfigurationExternal` can return a nil `gardenletConfig`. We should introduce a nil check.

I send you per PN my ManagedSeed resource that causes this to happen.

Reply:
Thanks, will take a look.

Reply:
I also ran into a panic here, but in my case `ManagedSeed.spec.gardenlet.config` was not `nil`.
```text
time="2021-02-10T10:22:27+01:00" level=info msg="Deploying gardenlet into shoot garden/seed-gcp2" managedSeed=garden/gcp
E0210 10:22:27.954075   24971 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 660 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2b3e620, 0x3fc7b40)
	/Users/d067603/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/Users/d067603/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x89
panic(0x2b3e620, 0x3fc7b40)
	/usr/local/Cellar/go/1.15.6/libexec/src/runtime/panic.go:969 +0x1b9
github.com/gardener/gardener/pkg/gardenlet/controller/managedseed.(*actuator).deployGardenlet(0xc000a7e240, 0x30e8ce0, 0xc0007b6c80, 0x3110940, 0xc0007630e0, 0xc0003371e0, 0xc00023e700, 0x0, 0x308c560)
	/Users/d067603/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/managedseed/managedseed_actuator.go:232 +0x75
github.com/gardener/gardener/pkg/gardenlet/controller/managedseed.(*actuator).Reconcile(0xc000a7e240, 0x30e8ce0, 0xc0007b6c80, 0xc0003371e0, 0xc00023e700, 0x2e01841, 0x4)
	/Users/d067603/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/managedseed/managedseed_actuator.go:107 +0x56a
github.com/gardener/gardener/pkg/gardenlet/controller/managedseed.(*reconciler).reconcile(0xc000a7e280, 0x30e8ce0, 0xc0007b6c80, 0xc0003371e0, 0xc0017041c0, 0xc001704100, 0x0, 0x0, 0x0)
	/Users/d067603/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/managedseed/managedseed_reconciler.go:113 +0x8a2
github.com/gardener/gardener/pkg/gardenlet/controller/managedseed.(*reconciler).Reconcile(0xc000a7e280, 0x30e8ce0, 0xc0007b6c80, 0xc00093c380, 0x6, 0xc00093c387, 0x3, 0x0, 0x72889f8, 0x70, ...)
	/Users/d067603/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/managedseed/managedseed_reconciler.go:74 +0x457
github.com/gardener/gardener/pkg/controllerutils.worker.func1(0x3106360, 0xc0005b1ea0, 0x308b4a0, 0xc000a7e280, 0x30e8ce0, 0xc0007b6c80, 0x2e07dab, 0xb, 0x4725100)
	/Users/d067603/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:114 +0x13e
github.com/gardener/gardener/pkg/controllerutils.worker(0x30e8ce0, 0xc0007b6c80, 0x3106360, 0xc0005b1ea0, 0x2e07dab, 0xb, 0x308b4a0, 0xc000a7e280)
	/Users/d067603/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:131 +0x7a
github.com/gardener/gardener/pkg/controllerutils.CreateWorker.func2.1(0x30e8ce0, 0xc0007b6c80)
	/Users/d067603/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:72 +0x6b
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/Users/d067603/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0014046e8)
	/Users/d067603/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00249bee8, 0x308c800, 0xc00002af60, 0xc0007b6c01, 0xc00095a3c0)
	/Users/d067603/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0014046e8, 0x3b9aca00, 0x0, 0x7622202020200a01, 0xc00095a3c0)
	/Users/d067603/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x30e8ce0, 0xc0007b6c80, 0xc001404748, 0x3b9aca00, 0x0, 0xc0007b4801)
	/Users/d067603/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(...)
	/Users/d067603/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:99
github.com/gardener/gardener/pkg/controllerutils.CreateWorker.func2(0x30e8ce0, 0xc0007b6c80, 0x3106360, 0xc0005b1ea0, 0x2e07dab, 0xb, 0x308b4a0, 0xc000a7e280, 0xc0007e6480, 0xc0004b0380)
	/Users/d067603/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:71 +0xd2
created by github.com/gardener/gardener/pkg/controllerutils.CreateWorker
	/Users/d067603/go/src/github.com/gardener/gardener/pkg/controllerutils/worker.go:70 +0x285
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x28db2b5]
```

```yaml
spec:
  gardenlet:
    bootstrap: ServiceAccount
    config:
      apiVersion: gardenlet.config.gardener.cloud/v1alpha1
      kind: GardenletConfiguration
      seedConfig:
        metadata:
          creationTimestamp: null
        spec:
          dns:
            provider:
              secretRef:
                name: seed-domain
                namespace: garden
              type: aws-route53
          ingress:
            controller:
              kind: nginx
            domain: ingress.gcp2.seed.[...]
          networks:
            nodes: 10.40.0.0/16
            pods: 10.41.128.0/17
            services: 10.41.0.0/17
          provider:
            region: europe-west1
            type: gcp
    deployment:
      image:
        pullPolicy: IfNotPresent
      replicaCount: 1
      revisionHistoryLimit: 1
      vpa: true
    mergeWithParent: true
  shoot:
    name: seed-gcp2
```

Reply:
I could not reproduce any issues with neither of your resources. On my test setup they simply reconciled successfully resulting in a working seed. 

@timebertt From your stack trace I assume the issue on your side is the same, for some reason the `gardenletConfig` returned by `helper.DecodeGardenletConfigurationExternal` is nil.

Let's perhaps debug this offline to avoid flooding the PR.

Reply:
OK, we looked at this together and figured out that the difference was due to the fact that i was using JSON, while you were using protobuf encoding when reading from the API server, and it seems the behavior of the two encoders was different with `RawExtension` fields. I added some conversion code to properly initialize the `Raw` field, it should be fixed now. Tested it on my side with protobuf and didn't see any more issues.

Reply:
We had an even more in-depth look into this now and I think, we finally figured out, how `runtime.RawExtension` is supposed to be used.
Long story short:
- `v1alpha1.ManagedSeed.spec.gardenlet.config` (external type) should be of type `runtime.RawExtension` instead of a pointer (`*runtime.RawExtensions`). This will auto generate conversion code between `runtime.Object` (internal version) and `runtime.RawExtension` (external) version and vice-versa.
- Normally, the API server doesn't know the concrete type of the `RawExtension` field and therefore decodes into `*runtime.Unknown`, which the generated conversion can handle (it sets `Unknown.Raw` to the raw JSON bytes when converting to internal and `RawExtension.Raw` when converting to external). This way, the raw bytes will be transported through all conversions and so on, which makes it possible for the protobuf serializer to encode the value again. While the json serializer is also able to encode the `RawExtension.Object` field, the protobuf serializer can only encode the `RawExtension.Raw` field (as raw bytes) because the concrete API type might not have any protobuf code (as it is the case for our `GardenletConfiguration`).
- in order to do API defaulting for the `ManagedSeed` object, we have to decode the object into a concrete `GardenletConfiguration` value (note: normally, the API server wouldn't do such things, it wouldn't act on the `RawExtension`'s content in any way). So this decoding, encoding in the defaulting logic is actually similar to use cases where the API server doesn't know about the concrete type and mutating webhooks are called for defaulting `RawExtension`s. The trick here is, to not set the `RawExtension.Object` field again after defaulting and only set `RawExtension.Raw`, as `.Object` will superseed `.Raw` on conversion, meaning the raw bytes will get lost again and the protobuf serializer won't be able to encode the config, resulting in `gardenlet.config=nil`.

We tested this approach and it is looking very promising, so @stoyanr wants to follow it.
He's currently doing the described changes and will push a new commit for it afterwards.
/author-action

Reply:
It turned out to be even more interesting than @timebertt wrote above. Relying on `Unknown` doesn't work because then the admission plugin can't simply convert, it needs to decode / encode again.

We had another look, and finally settled for an approach where the custom conversion code does encoding / decoding, and everything else works more or less as before.

Just pushed a new version amending @timebertt's commit to follow this approach. Was able to test successfully with it with both protobuf and JSON.